File: src/android/plugin/google/maps/PluginPolygon.java
package plugin.google.maps;

import com.google.android.libraries.maps.GoogleMap;
import com.google.android.libraries.maps.model.Circle;
import com.google.android.libraries.maps.model.LatLng;
import com.google.android.libraries.maps.model.LatLngBounds;
import com.google.android.libraries.maps.model.Polygon;
import com.google.android.libraries.maps.model.PolygonOptions;

import org.apache.cordova.CallbackContext;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

//Future implement
//TODO: https://codepen.io/jhawes/pen/ujdgK


public class PluginPolygon extends MyPlugin implements IOverlayPlugin {

    private PluginMap pluginMap;
    public final ConcurrentHashMap<String, MetaPolygon> objects = new ConcurrentHashMap<String, MetaPolygon>();


    public PluginMap getMapInstance(String mapId) {
        return (PluginMap) CordovaGoogleMaps.viewPlugins.get(mapId);
    }
    public PluginPolygon getInstance(String mapId) {
        PluginMap mapInstance = getMapInstance(mapId);
        return (PluginPolygon) mapInstance.plugins.get(String.format("%s-polygon", mapId));
    }

    @Override
    public void setPluginMap(PluginMap pluginMap) {
        this.pluginMap = pluginMap;
    }

    /**
     * Create polygon
     */
    public void create(final JSONArray args, final CallbackContext callbackContext) throws JSONException {

        final PolygonOptions polygonOptions = new PolygonOptions();
        int color;
        final LatLngBounds.Builder builder = new LatLngBounds.Builder();
        final JSONObject properties = new JSONObject();
        final ArrayList<LatLngBounds> holeBounds = new ArrayList<LatLngBounds>();
        final ArrayList<LatLng> path = new ArrayList<LatLng>();
        final ArrayList<ArrayList<LatLng>> holePaths = new ArrayList<ArrayList<LatLng>>();


        JSONObject opts = args.getJSONObject(2);
        final String hashCode = args.getString(3);

        final String polygonId = String.format("polygon_%s", hashCode);
        MetaPolygon meta = new MetaPolygon(polygonId);

        if (opts.has("points")) {
            JSONArray points = opts.getJSONArray("points");
            ArrayList<LatLng> path2 = PluginUtil.JSONArray2LatLngList(points);
            for (int i = 0; i < path2.size(); i++) {
                polygonOptions.add(path2.get(i));
                path.add(path2.get(i));
                builder.include(path2.get(i));
            }
        }

        if (opts.has("holes")) {
            JSONArray holes = opts.getJSONArray("holes");
            int i;
            JSONArray latLngArray;
            ArrayList<LatLng> hole;
            Iterator<LatLng> iterator;
            LatLng latLng;
            LatLngBounds.Builder builder2;
            for (i = 0; i < holes.length(); i++) {
                latLngArray = holes.getJSONArray(i);
                hole = PluginUtil.JSONArray2LatLngList(latLngArray);
                polygonOptions.addHole(hole);
                holePaths.add(hole);
                iterator = hole.iterator();
                builder2 = new LatLngBounds.Builder();
                while (iterator.hasNext()) {
                    latLng = iterator.next();
                    builder2.include(latLng);
                }
                holeBounds.add(builder2.build());
            }
        }
        if (opts.has("strokeColor")) {
            color = PluginUtil.parsePluginColor(opts.getJSONArray("strokeColor"));
            polygonOptions.strokeColor(color);
        }
        if (opts.has("fillColor")) {
            color = PluginUtil.parsePluginColor(opts.getJSONArray("fillColor"));
            polygonOptions.fillColor(color);
        }
        if (opts.has("strokeWidth")) {
            polygonOptions.strokeWidth((float)(opts.getDouble("strokeWidth") * density));
        }
        if (opts.has("visible")) {
            meta.isVisible = opts.getBoolean("visible");
            polygonOptions.visible(meta.isVisible);
        }
        if (opts.has("geodesic")) {
            polygonOptions.geodesic(opts.getBoolean("geodesic"));
        }
        if (opts.has("zIndex")) {
            polygonOptions.zIndex(opts.getInt("zIndex"));
        }
        if (opts.has("clickable")) {
            meta.isClickable = opts.getBoolean("clickable");
            properties.put("isClickable", meta.isClickable);
        } else {
            properties.put("isClickable", true);
            meta.isClickable = true;
        }
        properties.put("isVisible", polygonOptions.isVisible());
        properties.put("zIndex", polygonOptions.getZIndex());
        properties.put("isGeodesic", polygonOptions.isGeodesic());

        // Since this plugin uses own detecting process,
        // set false to the clickable property.
        polygonOptions.clickable(false);

        objects.put(polygonId, meta);

        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                meta.polygon = pluginMap.getGoogleMap().addPolygon(polygonOptions);
                meta.polygon.setTag(polygonId);
                meta.bounds = builder.build();
                meta.path = path;
                meta.holePaths = holePaths;
                meta.properties = properties;
            }
        });

        JSONObject result = new JSONObject();
        try {
            result.put("hashCode", hashCode);
            result.put("__pgmId", polygonId);
        } catch (JSONException e) {
            e.printStackTrace();
        }
        callbackContext.success(result);
    }


    @Override
    public void onDestroy() {
        super.onDestroy();
        objects.clear();
    }

    /**
     * set fill color
     */
    @PgmPluginMethod(runOnUiThread = true)
    public void setFillColor(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        PluginPolygon instance = getInstance(mapId);
        int color = PluginUtil.parsePluginColor(args.getJSONArray(2));
        Polygon polygon = instance.objects.get(polygonId).polygon;
        polygon.setFillColor(color);
        callbackContext.success();
    }

    /**
     * set stroke color
     */
    @PgmPluginMethod(runOnUiThread = true)
    public void setStrokeColor(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        PluginPolygon instance = getInstance(mapId);
        int color = PluginUtil.parsePluginColor(args.getJSONArray(2));
        Polygon polygon = instance.objects.get(polygonId).polygon;
        polygon.setStrokeColor(color);
        callbackContext.success();
    }

    /**
     * set stroke width
     */
    @PgmPluginMethod(runOnUiThread = true)
    public void setStrokeWidth(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        PluginPolygon instance = getInstance(mapId);
        float width = (float)(args.getDouble(2) * density);
        Polygon polygon = instance.objects.get(polygonId).polygon;
        polygon.setStrokeWidth(width);
        callbackContext.success();
    }

    /**
     * set z-index
     */
    @PgmPluginMethod(runOnUiThread = true)
    public void setZIndex(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        PluginPolygon instance = getInstance(mapId);
        float zIndex = (float) args.getDouble(2);
        Polygon polygon = instance.objects.get(polygonId).polygon;
        polygon.setZIndex(zIndex);
        callbackContext.success();
    }

    /**
     * set geodesic
     */
    @PgmPluginMethod(runOnUiThread = true)
    public void setGeodesic(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        PluginPolygon instance = getInstance(mapId);
        boolean isGeodisic = args.getBoolean(2);
        Polygon polygon = instance.objects.get(polygonId).polygon;
        polygon.setGeodesic(isGeodisic);
        callbackContext.success();
    }

    /**
     * Remove the polygon
     */
    @PgmPluginMethod(runOnUiThread = true)
    public void remove(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        PluginPolygon instance = getInstance(mapId);
        MetaPolygon meta = instance.objects.remove(polygonId);
        if (meta == null) {
            callbackContext.success();
            return;
        }
        Polygon polygon = meta.polygon;
        polygon.remove();
        callbackContext.success();
    }

    /**
     * Set points
     */
    @PgmPluginMethod(runOnUiThread = true)
    public void removePointAt(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        int index = args.getInt(2);
        PluginPolygon instance = getInstance(mapId);
        MetaPolygon meta = instance.objects.get(polygonId);
        Polygon polygon = meta.polygon;


        //------------------------
        // Update the hole list
        //------------------------
        String propertyId = "polygon_path_" + polygonId;
        final ArrayList<LatLng> path = meta.path;
        if (path.size() > 0) {
            path.remove(index);
        }

        //-----------------------------------
        // Recalculate the polygon bounds
        //-----------------------------------
        propertyId = "polygon_bounds_" + polygonId;
        if (path.size() > 0) {
            meta.bounds = PluginUtil.getBoundsFromPath(path);
        } else {
            meta.bounds = null;
        }

        if (path.size() > 0) {
            try {
                polygon.setPoints(path);
            } catch (Exception e) {
                // Ignore this error
                //e.printStackTrace();
            }
        } else {
            meta.isVisible = false;
            polygon.setVisible(false);
        }
        callbackContext.success();
    }

    /**
     * Set points
     */
    @PgmPluginMethod(runOnUiThread = true)
    public void setPoints(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        PluginPolygon instance = getInstance(mapId);
        MetaPolygon meta = instance.objects.get(polygonId);
        Polygon polygon = meta.polygon;

        JSONArray positionList = args.getJSONArray(2);

        //------------------------
        // Update the hole list
        //------------------------
        String propertyId = "polygon_path_" + polygonId;
        meta.path.clear();
        JSONObject position;
        for (int i = 0; i < positionList.length(); i++) {
            position = positionList.getJSONObject(i);
            meta.path.add(new LatLng(position.getDouble("lat"), position.getDouble("lng")));
        }

        //-----------------------------------
        // Recalculate the polygon bounds
        //-----------------------------------
        meta.bounds = PluginUtil.getBoundsFromPath(meta.path);

        // Update the polygon
        polygon.setPoints(meta.path);
        if (meta.path.size() > 0) {
            meta.isVisible = polygon.isVisible();
            polygon.setVisible(meta.isVisible);
        } else {meta.isVisible = false;
            polygon.setVisible(false);
        }
        callbackContext.success();
    }

    /**
     * Insert a point
     */
    @PgmPluginMethod(runOnUiThread = true)
    public void insertPointAt(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        PluginPolygon instance = getInstance(mapId);
        MetaPolygon meta = instance.objects.get(polygonId);

        int index = args.getInt(2);
        JSONObject position = args.getJSONObject(3);
        final LatLng latLng = new LatLng(position.getDouble("lat"), position.getDouble("lng"));

        //------------------------
        // Update the hole list
        //------------------------
        boolean shouldBeVisible = false;
        if (meta.path.size() == 0) {
            if (meta.properties.getBoolean("isVisible")) {
                shouldBeVisible = true;
            }
        }
        meta.path.add(index, latLng);

        //-----------------------------------
        // Recalculate the polygon bounds
        //-----------------------------------
        meta.bounds = PluginUtil.getBoundsFromPath(meta.path);

        boolean changeToVisible = shouldBeVisible;
        meta.polygon.setPoints(meta.path);
        if (changeToVisible) {
            meta.isVisible = true;
            meta.polygon.setVisible(true);
        }
        callbackContext.success();
    }

    /**
     * Set points
     */
    @PgmPluginMethod(runOnUiThread = true)
    public void setPointAt(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        PluginPolygon instance = getInstance(mapId);
        MetaPolygon meta = instance.objects.get(polygonId);

        int index = args.getInt(2);
        JSONObject position = args.getJSONObject(3);
        final LatLng latLng = new LatLng(position.getDouble("lat"), position.getDouble("lng"));

        //------------------------
        // Update the hole list
        //------------------------
        meta.path.set(index, latLng);

        //-----------------------------------
        // Recalculate the polygon bounds
        //-----------------------------------
        meta.bounds = PluginUtil.getBoundsFromPath(meta.path);

        // Update the polygon
        meta.polygon.setPoints(meta.path);
        callbackContext.success();
    }

    /**
     * Set points
     */
    @PgmPluginMethod(runOnUiThread = true)
    public void setHoles(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        PluginPolygon instance = getInstance(mapId);
        MetaPolygon meta = instance.objects.get(polygonId);

        JSONArray holeList = args.getJSONArray(2);

        //------------------------
        // Update the hole list
        //------------------------
        for (int i = 0; i < meta.holePaths.size(); i++) {
            meta.holePaths.get(i).clear();
        }
        meta.holePaths.clear();

        JSONObject position;
        for (int i = 0; i < holeList.length(); i++) {
            ArrayList<LatLng> hole = new ArrayList<LatLng>();
            JSONArray holePositions = holeList.getJSONArray(i);
            for (int j = 0; j < holePositions.length(); j++) {
                position = holePositions.getJSONObject(j);
                hole.add(new LatLng(position.getDouble("lat"), position.getDouble("lng")));
            }
            meta.holePaths.add(hole);
        }

        // Update the polygon
        meta.polygon.setHoles(meta.holePaths);
        callbackContext.success();
    }
    /**
     * Set points
     */
    @PgmPluginMethod(runOnUiThread = true)
    public void insertPointOfHoleAt(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        PluginPolygon instance = getInstance(mapId);
        MetaPolygon meta = instance.objects.get(polygonId);

        final int holeIndex = args.getInt(2);
        final int pointIndex = args.getInt(3);
        JSONObject position = args.getJSONObject(4);
        final LatLng latLng = new LatLng(position.getDouble("lat"), position.getDouble("lng"));


        //------------------------
        // Update the hole list
        //------------------------
        ArrayList<LatLng> hole = null;
        if (holeIndex < meta.holePaths.size()) {
            hole = meta.holePaths.get(holeIndex);
        }
        if (hole == null) {
            hole = new ArrayList<LatLng>();
        }
        if (meta.holePaths.size() == 0) {
            meta.holePaths.add(hole);
        }
        hole.add(pointIndex, latLng);

        // Update the polygon
        meta.polygon.setHoles(meta.holePaths);
        callbackContext.success();
    }

    /**
     * Set points
     */
    @PgmPluginMethod(runOnUiThread = true)
    public void setPointOfHoleAt(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        PluginPolygon instance = getInstance(mapId);
        MetaPolygon meta = instance.objects.get(polygonId);

        int holeIndex = args.getInt(2);
        int pointIndex = args.getInt(3);
        JSONObject position = args.getJSONObject(4);
        LatLng latLng = new LatLng(position.getDouble("lat"), position.getDouble("lng"));

        //------------------------
        // Update the hole list
        //------------------------
        ArrayList<LatLng> hole = null;
        if (holeIndex < meta.holePaths.size()) {
            hole = meta.holePaths.get(holeIndex);
        }
        if (hole == null) {
            hole = new ArrayList<LatLng>();
        }
        if (meta.holePaths.size() == 0) {
            meta.holePaths.add(hole);
        }
        hole.set(pointIndex, latLng);

        // Update the polygon
        meta.polygon.setHoles(meta.holePaths);
        callbackContext.success();
    }

    /**
     * Set points
     */
    @PgmPluginMethod(runOnUiThread = true)
    public void removePointOfHoleAt(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        PluginPolygon instance = getInstance(mapId);
        MetaPolygon meta = instance.objects.get(polygonId);

        final int holeIndex = args.getInt(2);
        final int pointIndex = args.getInt(3);

        //------------------------
        // Update the hole list
        //------------------------
        String propertyId = "polygon_holePaths_" + polygonId;
        ArrayList<LatLng> hole = null;
        if (holeIndex < meta.holePaths.size()) {
            hole = meta.holePaths.get(holeIndex);
        }
        if (hole == null) {
            hole = new ArrayList<LatLng>();
        }
        if (meta.holePaths.size() == 0) {
            meta.holePaths.add(hole);
        }
        hole.remove(pointIndex);

        // Update the polygon
        meta.polygon.setHoles(meta.holePaths);
        callbackContext.success();
    }


    @PgmPluginMethod(runOnUiThread = true)
    public void insertHoleAt(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        PluginPolygon instance = getInstance(mapId);
        MetaPolygon meta = instance.objects.get(polygonId);

        final int holeIndex = args.getInt(2);

        JSONArray holeJson = args.getJSONArray(3);
        final ArrayList<LatLng> newHole = PluginUtil.JSONArray2LatLngList(holeJson);

        //------------------------
        // Update the hole list
        //------------------------
        meta.holePaths.add(holeIndex, newHole);

        // Update the polygon
        meta.polygon.setHoles(meta.holePaths);
        callbackContext.success();
    }

    /**
     * Set points
     */
    @PgmPluginMethod(runOnUiThread = true)
    public void setHoleAt(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        PluginPolygon instance = getInstance(mapId);
        MetaPolygon meta = instance.objects.get(polygonId);

        //------------------------
        // Update the hole list
        //------------------------
        int holeIndex = args.getInt(2);
        JSONArray holeJson = args.getJSONArray(3);
        meta.holePaths.set(holeIndex, PluginUtil.JSONArray2LatLngList(holeJson));

        // Update the polygon
        meta.polygon.setHoles(meta.holePaths);
        callbackContext.success();
    }


    /**
     * Set points
     */
    @PgmPluginMethod(runOnUiThread = true)
    public void removeHoleAt(final JSONArray args, final CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        PluginPolygon instance = getInstance(mapId);
        MetaPolygon meta = instance.objects.get(polygonId);

        //------------------------
        // Update the hole list
        //------------------------
        int holeIndex = args.getInt(2);
        meta.holePaths.remove(holeIndex);

        // Update the polygon
        meta.polygon.setHoles(meta.holePaths);
        callbackContext.success();
    }
    /**
     * Set visibility for the object
     */
    @PgmPluginMethod(runOnUiThread = true)
    public void setVisible(JSONArray args, CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        PluginPolygon instance = getInstance(mapId);
        MetaPolygon meta = instance.objects.get(polygonId);

        boolean isVisible = args.getBoolean(2);
        meta.polygon.setVisible(isVisible);
        meta.isVisible = isVisible;

        meta.properties.put("isVisible", isVisible);
        callbackContext.success();
    }

    /**
     * Set clickable for the object
     */
    @PgmPluginMethod
    public void setClickable(JSONArray args, CallbackContext callbackContext) throws JSONException {
        String mapId = args.getString(0);
        String polygonId = args.getString(1);
        PluginPolygon instance = getInstance(mapId);
        MetaPolygon meta = instance.objects.get(polygonId);

        boolean clickable = args.getBoolean(2);
        meta.isClickable = clickable;
        meta.properties.put("isClickable", clickable);
        callbackContext.success();
    }
}

File: src/browser/PluginPolygon.js



var utils = require('cordova/utils'),
  event = require('cordova-plugin-googlemaps.event'),
  BaseClass = require('cordova-plugin-googlemaps.BaseClass'),
  LatLng = require('cordova-plugin-googlemaps.LatLng');

function PluginPolygon(pluginMap) {
  var self = this;
  BaseClass.apply(self);
  Object.defineProperty(self, "pluginMap", {
    value: pluginMap,
    writable: false
  });
}

utils.extend(PluginPolygon, BaseClass);

PluginPolygon.prototype._create = function(onSuccess, onError, args) {
  var self = this,
    map = self.pluginMap.get('map'),
    polygonId = 'polygon_' + args[2],
    pluginOptions = args[1];

  var polygonOpts = {
    'overlayId': polygonId,
    'map': map,
    'paths': new google.maps.MVCArray()
  };

  if (pluginOptions.points) {
    var strokePath = new google.maps.MVCArray();
    pluginOptions.points.forEach(function(point) {
      strokePath.push(new google.maps.LatLng(point.lat, point.lng));
    });
    polygonOpts.paths.push(strokePath);


    if (Array.isArray(pluginOptions.holes)) {
      pluginOptions.holes.forEach(function(hole) {
        var holeMvc = new google.maps.MVCArray();
        hole.forEach(function(vertix) {
          holeMvc.push(new google.maps.LatLng(vertix.lat, vertix.lng));
        });
        polygonOpts.paths.push(holeMvc);
      });
    }
  }
  if (Array.isArray(pluginOptions.strokeColor)) {
    polygonOpts.strokeColor = 'rgb(' + pluginOptions.strokeColor[0] + ',' + pluginOptions.strokeColor[1] + ',' + pluginOptions.strokeColor[2] + ')';
    polygonOpts.strokeOpacity = pluginOptions.strokeColor[3] / 256;
  }
  if (Array.isArray(pluginOptions.fillColor)) {
    polygonOpts.fillColor = 'rgb(' + pluginOptions.fillColor[0] + ',' + pluginOptions.fillColor[1] + ',' + pluginOptions.fillColor[2] + ')';
    polygonOpts.fillOpacity = pluginOptions.fillColor[3] / 256;
  }
  if ('width' in pluginOptions) {
    polygonOpts.strokeWeight = pluginOptions.width;
  }
  if ('zIndex' in pluginOptions) {
    polygonOpts.zIndex = pluginOptions.zIndex;
  }
  if ('visible' in pluginOptions) {
    polygonOpts.visible = pluginOptions.visible;
  }
  if ('geodesic' in pluginOptions) {
    polygonOpts.geodesic = pluginOptions.geodesic;
  }
  if ('clickable' in pluginOptions) {
    polygonOpts.clickable = pluginOptions.clickable;
  }

  var polygon = new google.maps.Polygon(polygonOpts);
  polygon.addListener('click', function(polyMouseEvt) {
    self._onPolygonEvent.call(self, polygon, polyMouseEvt);
  });

  self.pluginMap.objects[polygonId] = polygon;

  onSuccess({
    'id': polygonId
  });
};

PluginPolygon.prototype.setFillColor = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var fillColor = args[1];
  var polygon = self.pluginMap.objects[overlayId];
  if (polygon) {

    if (Array.isArray(fillColor)) {
      polygon.setOptions({
        'fillColor': 'rgb(' + fillColor[0] + ',' + fillColor[1] + ',' + fillColor[2] + ')',
        'fillOpacity': fillColor[3] / 256
      });
    }
  }
  onSuccess();
};

PluginPolygon.prototype.setStrokeColor = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var polygon = self.pluginMap.objects[overlayId];
  var strokeColor = args[1];
  if (polygon) {
    if (Array.isArray(strokeColor)) {
      polygon.setOptions({
        'strokeColor': 'rgb(' + strokeColor[0] + ',' + strokeColor[1] + ',' + strokeColor[2] + ')',
        'strokeOpacity': strokeColor[3] / 256
      });
    }
  }
  onSuccess();
};

PluginPolygon.prototype.setStrokeWidth = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var width = args[1];
  var polygon = self.pluginMap.objects[overlayId];
  if (polygon) {
    polygon.setOptions({
      'strokeWeight': width
    });
  }
  onSuccess();
};


PluginPolygon.prototype.setZIndex = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var zIndex = args[1];
  var polygon = self.pluginMap.objects[overlayId];
  if (polygon) {
    polygon.setOptions({
      'zIndex': zIndex
    });
  }
  onSuccess();
};

PluginPolygon.prototype.setPoints = function(onSuccess, onError, args) {
  var self = this,
    polygonId = args[0],
    positionList = args[1],
    polygon = self.pluginMap.objects[polygonId];
  if (polygon) {
    //------------------------
    // Update the points list
    //------------------------
    polygon.setPath(positionList);
  }
  onSuccess();
};

PluginPolygon.prototype.setClickable = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var clickable = args[1];
  var polygon = self.pluginMap.objects[overlayId];
  if (polygon) {
    polygon.setOptions({
      'clickable': clickable === true
    });
  }
  onSuccess();
};

PluginPolygon.prototype.setVisible = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var polygon = self.pluginMap.objects[overlayId];
  if (polygon) {
    polygon.setVisible(args[1]);
  }
  onSuccess();
};

PluginPolygon.prototype.setGeodesic = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var polygon = self.pluginMap.objects[overlayId];
  if (polygon) {
    polygon.setOptions({
      'geodesic': args[1] === true
    });
  }
  onSuccess();
};
PluginPolygon.prototype.insertPointAt = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var polygon = self.pluginMap.objects[overlayId];
  if (polygon) {
    var index = args[1];
    var latLng = new google.maps.LatLng(args[2].lat, args[2].lng);
    polygon.getPath().insertAt(index, latLng);
  }
  onSuccess();
};
PluginPolygon.prototype.setPointAt = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var polygon = self.pluginMap.objects[overlayId];
  if (polygon) {
    var index = args[1];
    var latLng = new google.maps.LatLng(args[2].lat, args[2].lng);
    polygon.getPath().setAt(index, latLng);
  }
  onSuccess();
};
PluginPolygon.prototype.removePointAt = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var polygon = self.pluginMap.objects[overlayId];
  if (polygon) {
    var index = args[1];
    polygon.getPath().removeAt(index, latLng);
  }
  onSuccess();
};

PluginPolygon.prototype.setHoles = function(onSuccess, onError, args) {
  var self = this,
    overlayId = args[0],
    polygon = self.pluginMap.objects[overlayId];
  if (polygon) {
    var paths = polygon.getPaths(),
      holeList = args[1],
      newPaths = new google.maps.MVCArray([paths.getAt(0)]);

    holeList.forEach(function(holeVertixes) {
      var hole = new google.maps.MVCArray();
      holeVertixes.forEach(function(vertix) {
        hole.push(new google.maps.LatLng(vertix.lat, vertix.lng));
      });
      newPaths.push(hole);
    });
    polygon.setPaths(newPaths);
  }
  onSuccess();
};

PluginPolygon.prototype.insertPointOfHoleAt = function(onSuccess, onError, args) {
  var self = this,
    overlayId = args[0],
    holeIndex = args[1] + 1,  // idx=0 is for outter vertixes
    pointIndex = args[2],
    position = args[3],
    polygon = self.pluginMap.objects[overlayId];

  if (polygon) {
    var index = args[1];
    var latLng = new google.maps.LatLng(position.lat, position.lng);
    var paths = polygon.getPaths();
    var hole = null;
    if (holeIndex < paths.getLength()) {
      hole = paths.getAt(holeIndex);
    }
    if (!hole) {
      hole = new google.maps.MVCArray();
      paths.push(hole);
    }
    hole.insertAt(pointIndex, latLng);
  }
  onSuccess();
};

PluginPolygon.prototype.setPointOfHoleAt = function(onSuccess, onError, args) {
  var self = this,
    overlayId = args[0],
    holeIndex = args[1] + 1,
    pointIndex = args[2],
    position = args[3],
    polygon = self.pluginMap.objects[overlayId];

  if (polygon) {
    var index = args[1];
    var latLng = new google.maps.LatLng(position.lat, position.lng);
    polygon.getPaths().getAt(holeIndex).setAt(pointIndex, latLng);
  }
  onSuccess();
};

PluginPolygon.prototype.removePointOfHoleAt = function(onSuccess, onError, args) {
  var self = this,
    overlayId = args[0],
    holeIndex = args[1] + 1,
    pointIndex = args[2],
    polygon = self.pluginMap.objects[overlayId];

  if (polygon) {
    polygon.getPaths().getAt(holeIndex).removeAt(pointIndex);
  }
  onSuccess();
};


PluginPolygon.prototype.insertHoleAt = function(onSuccess, onError, args) {
  var self = this,
    overlayId = args[0],
    holeIndex = args[1] + 1,
    vertixes = args[2],
    polygon = self.pluginMap.objects[overlayId];

  if (polygon) {
    var hole = new google.maps.MVCArray();
    vertixes.forEach(function(vertix) {
      hole.push(new google.maps.LatLng(vertix.lat, vertix.lng));
    });
    polygon.getPaths().insertAt(holeIndex, hole);
  }
  onSuccess();
};



PluginPolygon.prototype.setHoleAt = function(onSuccess, onError, args) {
  var self = this,
    overlayId = args[0],
    holeIndex = args[1] + 1,
    vertixes = args[2],
    polygon = self.pluginMap.objects[overlayId];

  if (polygon) {
    var hole = new google.maps.MVCArray();
    vertixes.forEach(function(vertix) {
      hole.push(new google.maps.LatLng(vertix.lat, vertix.lng));
    });
    polygon.getPaths().setAt(holeIndex, hole);
  }
  onSuccess();
};


PluginPolygon.prototype.removeHoleAt = function(onSuccess, onError, args) {
  var self = this,
    overlayId = args[0],
    holeIndex = args[1] + 1,
    polygon = self.pluginMap.objects[overlayId];

  if (polygon) {
    polygon.getPaths().removeAt(holeIndex);
  }
  onSuccess();
};
PluginPolygon.prototype.remove = function(onSuccess, onError, args) {
  var self = this;
  var overlayId = args[0];
  var polygon = self.pluginMap.objects[overlayId];
  if (polygon) {
    google.maps.event.clearInstanceListeners(polygon);
    polygon.setMap(null);
    polygon = undefined;
    self.pluginMap.objects[overlayId] = undefined;
    delete self.pluginMap.objects[overlayId];
  }
  onSuccess();
};

PluginPolygon.prototype._onPolygonEvent = function(polygon, polyMouseEvt) {
  var self = this,
    mapId = self.pluginMap.id;
  if (mapId in plugin.google.maps) {
    plugin.google.maps[mapId]({
      'evtName': event.POLYGON_CLICK,
      'callback': '_onOverlayEvent',
      'args': [polygon.overlayId, new plugin.google.maps.LatLng(polyMouseEvt.latLng.lat(), polyMouseEvt.latLng.lng())]
    });
  }

};
module.exports = PluginPolygon;
